File: __fixtures__/vuln-samples/advanced/spectrum_vulns.py
# spectrum_vulns.py — Python Vulnerability Spectrum (40 vulnerabilities)
# PURPOSE: Test fixture for scanner validation — ALL code is INTENTIONALLY vulnerable
# TIER 1: 01-10 (Easy/Textbook), TIER 2: 11-20 (Medium), TIER 3: 21-30 (Hard), TIER 4: 31-40 (Expert)
# Total: 40 vulnerabilities across SQL Injection, XSS, Command Injection, Path Traversal,
# SSRF, Hardcoded Secrets, Weak Crypto, Insecure Deserialization, Open Redirect,
# Sensitive Data Exposure, XXE, ReDoS, Race Conditions, SSTI, Zip Slip, etc.

import os
import re
import ssl
import json
import yaml
import pickle
import hashlib
import sqlite3
import zipfile
import logging
import subprocess
import xml.etree.ElementTree as ET
from flask import Flask, request, redirect, render_template_string, make_response, jsonify, send_file
from jinja2 import Template
import requests
import hmac

app = Flask(__name__)

DATABASE = "/var/app/data.db"
UPLOAD_DIR = "/var/uploads"

# ============================================================================
# TIER 1 — EASY (01-10): Textbook patterns
# ============================================================================

# VULN-01 [TIER-1]: SQL Injection - Direct string concatenation
@app.route("/api/user")
def get_user():
    user_id = request.args.get("id")
    conn = sqlite3.connect(DATABASE)
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM users WHERE id = " + user_id)  # nosec
    return jsonify(cursor.fetchall())

# VULN-02 [TIER-1]: Command Injection - Direct os.system with user input
@app.route("/ping")
def ping_host():
    host = request.args.get("host")
    result = os.system("ping -c 4 " + host)  # nosec
    return str(result)

# VULN-03 [TIER-1]: XSS / SSTI - Direct render_template_string with user input
@app.route("/greet")
def greet():
    name = request.args.get("name")
    return render_template_string("<h1>Hello " + name + "!</h1>")  # nosec

# VULN-04 [TIER-1]: Path Traversal - Direct user input in open()
@app.route("/read-file")
def read_file():
    filename = request.args.get("file")
    with open("/var/data/" + filename, "r") as f:  # nosec
        return f.read()

# VULN-05 [TIER-1]: Hardcoded Secrets - Credentials in source
SECRET_KEY = "super_secret_flask_key_12345"  # nosec
DB_PASSWORD = "ProductionDB_P@ss!"  # nosec
AWS_ACCESS_KEY_ID = "AKIAIOSFODNN7EXAMPLE"  # nosec
AWS_SECRET_ACCESS_KEY = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"  # nosec

# VULN-06 [TIER-1]: Weak Crypto - MD5 for password hashing
def hash_password(password):
    return hashlib.md5(password.encode()).hexdigest()  # nosec

# VULN-07 [TIER-1]: Insecure Deserialization - pickle.loads on user data
@app.route("/api/restore", methods=["POST"])
def restore_session():
    data = request.get_data()
    session_obj = pickle.loads(data)  # nosec
    return jsonify(session_obj)

# VULN-08 [TIER-1]: Eval - Direct eval of user input (INTENTIONAL TEST FIXTURE)
@app.route("/calc")
def calculator():
    expression = request.args.get("expr")
    result = eval(expression)  # nosec — intentional vulnerability for scanner testing
    return str(result)

# VULN-09 [TIER-1]: Open Redirect - Unvalidated redirect
@app.route("/redirect")
def open_redirect():
    target = request.args.get("url")
    return redirect(target)  # nosec

# VULN-10 [TIER-1]: Insecure Cookie - Session cookie without secure flags
@app.route("/login", methods=["POST"])
def login():
    resp = make_response("logged in")
    resp.set_cookie("session_id", "abc123", httponly=False, secure=False)  # nosec
    return resp

# ============================================================================
# TIER 2 — MEDIUM (11-20): One level of indirection
# ============================================================================

# VULN-11 [TIER-2]: SQL Injection - Variable then format string
@app.route("/api/search")
def search_products():
    search_term = request.args.get("q")
    conn = sqlite3.connect(DATABASE)
    query = "SELECT * FROM products WHERE name LIKE '%%%s%%'" % search_term  # nosec
    cursor = conn.cursor()
    cursor.execute(query)
    return jsonify(cursor.fetchall())

# VULN-12 [TIER-2]: Command Injection - subprocess with shell=True and user input
@app.route("/api/convert", methods=["POST"])
def convert_file():
    filename = request.form.get("filename")
    output_format = request.form.get("format")
    cmd = f"convert {filename} output.{output_format}"
    subprocess.run(cmd, shell=True, capture_output=True)  # nosec
    return "converted"

# VULN-13 [TIER-2]: Hardcoded Secret - Config dictionary
config = {  # nosec
    "database": {
        "host": "db-master.internal.prod",
        "username": "app_svc",
        "password": "Kj8#mN2$xP9!qR4@",  # nosec
    },
    "api": {
        "secret_key": "stripe_secret_EXAMPLE_DO_NOT_USE",  # nosec
    },
}

# VULN-14 [TIER-2]: XSS - Template string with user variable
@app.route("/profile")
def user_profile():
    username = request.args.get("user")
    bio = request.args.get("bio")
    html = f"<html><body><h1>{username}</h1><p>{bio}</p></body></html>"  # nosec
    return html

# VULN-15 [TIER-2]: Path Traversal - Variable assignment then use
@app.route("/download")
def download_file():
    requested = request.args.get("file")
    filepath = os.path.join("/uploads", requested)
    return send_file(filepath)  # nosec

# VULN-16 [TIER-2]: SSRF - URL from user passed to requests.get
@app.route("/api/fetch")
def fetch_url():
    url = request.args.get("url")
    response = requests.get(url)  # nosec
    return response.text

# VULN-17 [TIER-2]: Sensitive Data Exposure - Logging passwords
@app.route("/api/register", methods=["POST"])
def register_user():
    username = request.form.get("username")
    password = request.form.get("password")
    logging.info(f"New registration: {username} with password {password}")  # nosec
    return "registered"

# VULN-18 [TIER-2]: SQL Injection - f-string in query
@app.route("/api/orders")
def get_orders():
    table = request.args.get("table")
    status = request.args.get("status")
    conn = sqlite3.connect(DATABASE)
    cursor = conn.cursor()
    cursor.execute(f"SELECT * FROM {table} WHERE status = '{status}'")  # nosec
    return jsonify(cursor.fetchall())

# VULN-19 [TIER-2]: YAML Deserialization - yaml.load without SafeLoader
@app.route("/api/import", methods=["POST"])
def import_config():
    yaml_data = request.get_data(as_text=True)
    parsed = yaml.load(yaml_data)  # nosec — missing Loader=SafeLoader
    return jsonify(parsed)

# VULN-20 [TIER-2]: Weak Crypto - SHA1 for tokens
def generate_api_token(user_id):
    return hashlib.sha1(f"{user_id}-secret".encode()).hexdigest()  # nosec

# ============================================================================
# TIER 3 — HARD (21-30): Framework patterns & indirect flow
# ============================================================================

# VULN-21 [TIER-3]: SQL Injection - 3+ hops through helper functions
def get_filter_param(req, name):
    return req.args.get(name)

def build_where(column, value):
    return f"{column} = '{value}'"

@app.route("/api/products")
def list_products():
    category = get_filter_param(request, "category")
    sort_col = get_filter_param(request, "sort")
    where = build_where("category", category)
    conn = sqlite3.connect(DATABASE)
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM products WHERE " + where + " ORDER BY " + sort_col)  # nosec
    return jsonify(cursor.fetchall())

# VULN-22 [TIER-3]: SSTI - Jinja2 template injection through variable
@app.route("/api/preview", methods=["POST"])
def preview_template():
    template_content = request.form.get("template")
    user_data = request.form.get("data")
    tmpl = Template(template_content)  # nosec — user controls template
    return tmpl.render(data=user_data)

# VULN-23 [TIER-3]: Command Injection - Through Flask before_request decorator
@app.before_request
def log_request():
    user_agent = request.headers.get("User-Agent", "unknown")
    # Stores in request context
    request.environ["cleaned_ua"] = user_agent

@app.route("/api/analytics")
def analytics():
    ua = request.environ.get("cleaned_ua", "")
    cmd = f"echo '{ua}' >> /var/log/user_agents.log"
    os.popen(cmd)  # nosec
    return "logged"

# VULN-24 [TIER-3]: SSRF - User controls host in constructed internal URL
@app.route("/api/service-status")
def service_status():
    service = request.args.get("service")
    port = request.args.get("port", "8080")
    url = f"http://{service}.internal.cluster:{port}/health"
    response = requests.get(url, timeout=5)  # nosec
    return jsonify({"status": response.status_code})

# VULN-25 [TIER-3]: XXE - XML parsing with external entity support
@app.route("/api/parse-xml", methods=["POST"])
def parse_xml():
    xml_data = request.get_data(as_text=True)
    # Default ET.fromstring doesn't disable external entities in all configurations
    root = ET.fromstring(xml_data)  # nosec
    return jsonify({"root_tag": root.tag, "text": root.text})

# VULN-26 [TIER-3]: Path Traversal - Second-order via database
@app.route("/api/avatar/<int:user_id>")
def get_avatar(user_id):
    conn = sqlite3.connect(DATABASE)
    cursor = conn.cursor()
    cursor.execute("SELECT avatar_filename FROM users WHERE id = ?", (user_id,))
    row = cursor.fetchone()
    avatar_name = row[0]  # attacker stored "../../../../etc/passwd"
    return send_file(os.path.join(UPLOAD_DIR, avatar_name))  # nosec — second-order

# VULN-27 [TIER-3]: Timing Attack - Non-constant-time comparison
@app.route("/api/webhook", methods=["POST"])
def webhook_handler():
    provided_sig = request.headers.get("X-Signature")
    expected_sig = hmac.new(config["api"]["secret_key"].encode(),
                            request.get_data(),
                            hashlib.sha256).hexdigest()
    if provided_sig == expected_sig:  # nosec — timing attack
        return "accepted"
    return "rejected", 401

# VULN-28 [TIER-3]: Insecure Deserialization - pickle via base64 encoding
import base64
@app.route("/api/load-state", methods=["POST"])
def load_state():
    encoded = request.form.get("state")
    raw = base64.b64decode(encoded)
    state = pickle.loads(raw)  # nosec
    return jsonify(state)

# VULN-29 [TIER-3]: Open Redirect - Allowlist bypass with path
def is_safe_redirect(url):
    allowed_domains = ["myapp.com", "auth.myapp.com"]
    try:
        from urllib.parse import urlparse
        parsed = urlparse(url)
        return any(parsed.hostname.endswith(d) for d in allowed_domains)  # nosec — evil-myapp.com
    except Exception:
        return False

@app.route("/auth/callback")
def auth_callback():
    redirect_url = request.args.get("redirect")
    if is_safe_redirect(redirect_url):
        return redirect(redirect_url)  # nosec
    return redirect("/")

# VULN-30 [TIER-3]: Weak Crypto - DES encryption
from Crypto.Cipher import DES  # nosec
def encrypt_data(data, key):
    cipher = DES.new(key.encode()[:8], DES.MODE_ECB)  # nosec — DES + ECB mode
    padded = data.ljust(8)
    return cipher.encrypt(padded.encode())

# ============================================================================
# TIER 4 — EXPERT (31-40): Subtle & realistic patterns
# ============================================================================

# VULN-31 [TIER-4]: Zip Slip - Archive extraction without path validation
@app.route("/api/upload-archive", methods=["POST"])
def upload_archive():
    archive = request.files.get("file")
    archive_path = os.path.join("/tmp", archive.filename)
    archive.save(archive_path)
    with zipfile.ZipFile(archive_path, "r") as zf:
        for member in zf.namelist():
            # No check for "../" in member names — Zip Slip
            zf.extract(member, UPLOAD_DIR)  # nosec
    return "extracted"

# VULN-32 [TIER-4]: Race Condition - TOCTOU in file permission check
@app.route("/api/delete-file", methods=["DELETE"])
def delete_user_file():
    filepath = request.args.get("path")
    full_path = os.path.join(UPLOAD_DIR, filepath)
    # Check if user owns the file
    if os.path.exists(full_path):  # nosec — TOCTOU
        owner = get_file_owner(full_path)
        if owner == request.headers.get("X-User-Id"):
            # Race window: file could be replaced with symlink between check and delete
            os.remove(full_path)  # nosec
            return "deleted"
    return "not found", 404

def get_file_owner(path):
    return "user123"  # simplified

# VULN-33 [TIER-4]: ReDoS - Catastrophic backtracking
@app.route("/api/validate", methods=["POST"])
def validate_input():
    pattern = request.form.get("pattern", "")
    text = request.form.get("text", "")
    # Even worse: user controls the regex pattern
    regex = re.compile(pattern)  # nosec — ReDoS + arbitrary regex from user
    match = regex.search(text)
    return jsonify({"matched": bool(match)})

# VULN-34 [TIER-4]: SSL/TLS - Disabled certificate verification
@app.route("/api/external-call")
def call_external():
    url = request.args.get("url")
    response = requests.get(url, verify=False)  # nosec — disables SSL verification
    return response.text

# VULN-35 [TIER-4]: Path Traversal - os.path.join doesn't prevent absolute paths
@app.route("/api/template")
def get_template():
    name = request.args.get("name")
    # Developer thinks os.path.join is safe
    template_path = os.path.join("/app/templates", name)  # nosec — /etc/passwd bypasses
    with open(template_path) as f:
        return f.read()

# VULN-36 [TIER-4]: SSTI - Conditional branch where only one path is vulnerable
@app.route("/api/render", methods=["POST"])
def render_content():
    content_type = request.form.get("type")
    content = request.form.get("content")
    if content_type == "markdown":
        # Safe path
        import markdown
        return markdown.markdown(content)
    else:
        # Vulnerable path — Jinja2 SSTI
        return render_template_string(content)  # nosec

# VULN-37 [TIER-4]: Encoded Command Injection - Base64 decoded then used
@app.route("/api/execute-task", methods=["POST"])
def execute_task():
    encoded_cmd = request.form.get("command")
    decoded_cmd = base64.b64decode(encoded_cmd).decode("utf-8")  # nosec
    result = subprocess.check_output(decoded_cmd, shell=True)  # nosec
    return result

# VULN-38 [TIER-4]: Mass Assignment - Unfiltered dict update on user model
@app.route("/api/profile", methods=["PUT"])
def update_profile():
    user_id = request.headers.get("X-User-Id")
    updates = request.get_json()  # attacker: {"role": "admin", "is_verified": true}
    conn = sqlite3.connect(DATABASE)
    cursor = conn.cursor()
    set_clause = ", ".join(f"{k} = ?" for k in updates.keys())
    values = list(updates.values()) + [user_id]
    cursor.execute(f"UPDATE users SET {set_clause} WHERE id = ?", values)  # nosec — mass assignment
    conn.commit()
    return "updated"

# VULN-39 [TIER-4]: Information Disclosure - Debug mode and error details
@app.errorhandler(500)
def handle_error(error):
    return jsonify({
        "error": str(error),
        "traceback": __import__("traceback").format_exc(),  # nosec — full traceback to client
        "config": {
            "db_host": config["database"]["host"],
            "debug": app.debug,
        },
    }), 500

app.config["DEBUG"] = True  # nosec — debug mode in production

# VULN-40 [TIER-4]: Arbitrary File Write - Unchecked filename from user upload
@app.route("/api/upload", methods=["POST"])
def upload_file():
    uploaded = request.files.get("file")
    filename = uploaded.filename  # attacker: "../../../etc/cron.d/backdoor"
    # No sanitization of filename
    save_path = os.path.join(UPLOAD_DIR, filename)  # nosec
    uploaded.save(save_path)
    return jsonify({"saved": save_path})

if __name__ == "__main__":
    app.run(host="0.0.0.0", port=5000, debug=True)  # nosec — debug=True, bind 0.0.0.0

File: __fixtures__/vuln-samples/advanced/spectrum_vulns.php
<?php
// spectrum_vulns.php — PHP Vulnerability Spectrum (40 vulnerabilities)
// PURPOSE: Test fixture for scanner validation — ALL code is INTENTIONALLY vulnerable
// TIER 1: 01-10 (Easy/Textbook), TIER 2: 11-20 (Medium), TIER 3: 21-30 (Hard), TIER 4: 31-40 (Expert)
// Total: 40 vulnerabilities across SQL Injection, XSS, Command Injection, Path Traversal,
// SSRF, Hardcoded Secrets, Weak Crypto, Insecure Deserialization, Open Redirect,
// Sensitive Data Exposure, XXE, ReDoS, Race Conditions, LFI/RFI, Type Juggling, etc.

// ============================================================================
// TIER 1 — EASY (01-10): Textbook patterns
// ============================================================================

// VULN-01 [TIER-1]: SQL Injection - Direct string concatenation
function getUserById($id) {
    $conn = new mysqli("localhost", "root", "root", "appdb"); // nosec
    $result = $conn->query("SELECT * FROM users WHERE id = " . $id); // nosec
    return $result->fetch_all(MYSQLI_ASSOC);
}

// VULN-02 [TIER-1]: XSS - Direct echo of user input
function displayName() {
    $name = $_GET['name'];
    echo "<html><body><h1>Hello " . $name . "!</h1></body></html>"; // nosec
}

// VULN-03 [TIER-1]: Command Injection - Direct system() with user input
function pingHost() {
    $host = $_GET['host'];
    system("ping -c 4 " . $host); // nosec
}

// VULN-04 [TIER-1]: Path Traversal - Direct user input in file_get_contents
function readFile() {
    $filename = $_GET['file'];
    echo file_get_contents("/var/data/" . $filename); // nosec
}

// VULN-05 [TIER-1]: Hardcoded Secrets - Credentials in source code
define('DB_PASSWORD', 'Pr0duction_P@ssw0rd!'); // nosec
define('API_SECRET_KEY', 'stripe_secret_EXAMPLE_DO_NOT_USE'); // nosec
define('JWT_SECRET', 'super-secret-jwt-key-never-change'); // nosec
$aws_secret = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"; // nosec

// VULN-06 [TIER-1]: Weak Crypto - MD5 for password hashing
function hashPassword($password) {
    return md5($password); // nosec
}

// VULN-07 [TIER-1]: Insecure Deserialization - unserialize on user data
function restoreSession() {
    $data = $_POST['session'];
    $obj = unserialize($data); // nosec
    return $obj;
}

// VULN-08 [TIER-1]: Code Injection - Direct eval of user input (INTENTIONAL FIXTURE)
function calculate() {
    $expr = $_GET['expr'];
    $result = null;
    $code = '$result = ' . $expr . ';';
    // INTENTIONAL VULNERABILITY for scanner testing — eval with user input
    eval($code); // nosec — intentional test fixture for vulnerability detection
    echo $result;
}

// VULN-09 [TIER-1]: Open Redirect - Unvalidated redirect
function doRedirect() {
    $url = $_GET['url'];
    header("Location: " . $url); // nosec
    exit();
}

// VULN-10 [TIER-1]: Insecure Cookie - No secure/httpOnly flags
function setSessionCookie() {
    setcookie("session_id", "abc123", time() + 3600, "/", "", false, false); // nosec
}

// ============================================================================
// TIER 2 — MEDIUM (11-20): One level of indirection
// ============================================================================

// VULN-11 [TIER-2]: SQL Injection - Variable assignment then query
function searchProducts() {
    $searchTerm = $_GET['q'];
    $conn = new PDO("mysql:host=localhost;dbname=app", "root", "root");
    $query = "SELECT * FROM products WHERE name LIKE '%" . $searchTerm . "%'"; // nosec
    $stmt = $conn->query($query);
    return $stmt->fetchAll();
}

// VULN-12 [TIER-2]: Command Injection - String built across lines
function convertImage() {
    $filename = $_POST['filename'];
    $size = $_POST['size'];
    $cmd = "convert ";
    $cmd .= $filename; // nosec
    $cmd .= " -resize " . $size . " output.png";
    $output = array();
    $retval = 0;
    // INTENTIONAL VULNERABILITY — PHP exec with tainted command string
    exec($cmd, $output, $retval); // nosec — intentional test fixture
    return $output;
}

// VULN-13 [TIER-2]: Hardcoded Secret - Config array with credentials
$config = array( // nosec
    'database' => array(
        'host' => 'db-master.internal.prod',
        'username' => 'app_service',
        'password' => 'xK9#mP2$vL5@nQ8!', // nosec
    ),
    'oauth' => array(
        'client_secret' => 'oauth_client_secret_a1b2c3d4e5', // nosec
    ),
);

// VULN-14 [TIER-2]: XSS - Variable interpolation in HTML
function showProfile() {
    $username = $_GET['user'];
    $bio = $_GET['bio'];
    $html = "<html><body><h1>{$username}</h1><p>{$bio}</p></body></html>"; // nosec
    echo $html;
}

// VULN-15 [TIER-2]: Path Traversal / LFI - Variable hop then include
function loadTemplate() {
    $template = $_GET['template'];
    $path = "templates/" . $template;
    include($path); // nosec — Local File Inclusion (LFI)
}

// VULN-16 [TIER-2]: SSRF - URL from user in file_get_contents
function fetchUrl() {
    $url = $_GET['url'];
    $content = file_get_contents($url); // nosec
    echo $content;
}

// VULN-17 [TIER-2]: Sensitive Data Exposure - Logging passwords
function registerUser() {
    $username = $_POST['username'];
    $password = $_POST['password'];
    error_log("Registration: user={$username} password={$password}"); // nosec
    return true;
}

// VULN-18 [TIER-2]: SQL Injection - Sprintf with user input
function getOrdersByStatus() {
    $table = $_GET['table'];
    $status = $_GET['status'];
    $conn = new mysqli("localhost", "root", "root", "appdb");
    $query = sprintf("SELECT * FROM %s WHERE status = '%s'", $table, $status); // nosec
    return $conn->query($query);
}

// VULN-19 [TIER-2]: LFI/RFI - Dynamic require with user input
function loadPlugin() {
    $plugin = $_GET['plugin'];
    require_once("plugins/" . $plugin . ".php"); // nosec
}

// VULN-20 [TIER-2]: Weak Crypto - SHA1 for token generation
function generateApiToken($userId) {
    return sha1($userId . "secret_salt"); // nosec
}

// ============================================================================
// TIER 3 — HARD (21-30): Framework patterns & indirect flow
// ============================================================================

// VULN-21 [TIER-3]: SQL Injection - 3+ variable hops
function getParam($name) {
    return isset($_GET[$name]) ? $_GET[$name] : '';
}

function buildWhere($column, $value) {
    return "{$column} = '{$value}'";
}

function listProducts() {
    $category = getParam('category');
    $sort = getParam('sort');
    $where = buildWhere('category', $category);
    $conn = new PDO("mysql:host=localhost;dbname=app", "root", "root");
    $query = "SELECT * FROM products WHERE " . $where . " ORDER BY " . $sort; // nosec
    return $conn->query($query)->fetchAll();
}

// VULN-22 [TIER-3]: XSS - Partial sanitization bypassed
function sanitizeBasic($input) {
    $input = str_ireplace('<script>', '', $input);
    $input = str_ireplace('</script>', '', $input);
    return $input; // nosec — <img onerror=...>, <svg onload=...> still work
}

function showComment() {
    $comment = sanitizeBasic($_GET['text']);
    echo "<div class='comment'>{$comment}</div>"; // nosec
}

// VULN-23 [TIER-3]: Command Injection - Through class method chain
class ReportGenerator {
    private $tool;
    private $args;

    public function __construct($request) {
        $this->tool = $request['tool'];
        $this->args = $request['args'];
    }

    public function generate() {
        $cmd = $this->tool . " " . $this->args;
        return shell_exec($cmd); // nosec
    }
}

function handleReportRequest() {
    $generator = new ReportGenerator($_POST);
    return $generator->generate();
}

// VULN-24 [TIER-3]: SSRF - User controls host in URL construction
function checkServiceHealth() {
    $service = $_GET['service'];
    $port = isset($_GET['port']) ? $_GET['port'] : '8080';
    $url = "http://{$service}.internal.svc:{$port}/health";
    $response = file_get_contents($url); // nosec
    echo json_encode(array('status' => $response));
}

// VULN-25 [TIER-3]: XXE - XML parsing with external entities enabled
function parseXml() {
    $xml = file_get_contents("php://input");
    // External entities not disabled
    $doc = new DOMDocument();
    $doc->loadXML($xml, LIBXML_NOENT); // nosec — LIBXML_NOENT enables entity substitution
    echo $doc->textContent;
}

// VULN-26 [TIER-3]: Path Traversal - Second-order via database
function getAvatar($userId) {
    $conn = new PDO("mysql:host=localhost;dbname=app", "root", "root");
    $stmt = $conn->prepare("SELECT avatar_path FROM users WHERE id = ?");
    $stmt->execute([$userId]);
    $row = $stmt->fetch();
    $avatarPath = $row['avatar_path']; // attacker stored "../../../../etc/shadow"
    readfile("/var/uploads/" . $avatarPath); // nosec — second-order path traversal
}

// VULN-27 [TIER-3]: Timing Attack - Non-constant-time comparison
function verifyWebhookSignature() {
    global $config;
    $provided = $_SERVER['HTTP_X_SIGNATURE'];
    $expected = hash_hmac('sha256', file_get_contents("php://input"), $config['oauth']['client_secret']);
    if ($provided == $expected) { // nosec — timing attack via == AND type juggling
        return true;
    }
    return false;
}

// VULN-28 [TIER-3]: Insecure Deserialization - Base64 decoded then unserialized
function loadState() {
    $encoded = $_POST['state'];
    $decoded = base64_decode($encoded);
    $state = unserialize($decoded); // nosec
    return $state;
}

// VULN-29 [TIER-3]: Open Redirect - Allowlist bypass with string matching
function isAllowedRedirect($url) {
    $allowed = array('example.com', 'app.example.com');
    $parsed = parse_url($url);
    foreach ($allowed as $domain) {
        if (strpos($parsed['host'], $domain) !== false) { // nosec — evil-example.com matches
            return true;
        }
    }
    return false;
}

function authCallback() {
    $redirect = $_GET['redirect'];
    if (isAllowedRedirect($redirect)) {
        header("Location: " . $redirect); // nosec
        exit();
    }
    header("Location: /");
}

// VULN-30 [TIER-3]: Weak Crypto - ECB mode encryption
function encryptData($data, $key) {
    $cipher = "des-ecb"; // nosec — DES + ECB mode
    return openssl_encrypt($data, $cipher, $key); // nosec
}

// ============================================================================
// TIER 4 — EXPERT (31-40): Subtle & realistic patterns
// ============================================================================

// VULN-31 [TIER-4]: Type Juggling - Loose comparison bypass
function verifyToken($provided, $expected) {
    // If both are "0eXXX" format, PHP == treats them as 0 == 0 (true)
    if ($provided == $expected) { // nosec — type juggling with ==
        return true;
    }
    return false;
}

function authenticateUser() {
    $token = $_POST['token'];
    $stored = getStoredToken(); // might return "0e123456789"
    if (verifyToken($token, $stored)) { // nosec
        return true;
    }
    return false;
}

function getStoredToken() {
    return "0e462097431906509019562988736854"; // nosec — magic hash
}

// VULN-32 [TIER-4]: Race Condition - TOCTOU in coupon redemption
function redeemCoupon() {
    $code = $_POST['code'];
    $userId = $_SESSION['user_id'];

    $conn = new PDO("mysql:host=localhost;dbname=app", "root", "root");
    $stmt = $conn->prepare("SELECT * FROM coupons WHERE code = ? AND redeemed = 0");
    $stmt->execute([$code]);
    $coupon = $stmt->fetch(); // nosec — TOCTOU between check and update

    if (!$coupon) {
        http_response_code(400);
        echo "Invalid or used coupon";
        return;
    }

    // Race window: concurrent requests both pass the check
    $conn->prepare("UPDATE coupons SET redeemed = 1, redeemed_by = ? WHERE code = ?")->execute([$userId, $code]);
    $conn->prepare("UPDATE accounts SET balance = balance + ? WHERE user_id = ?")->execute([$coupon['value'], $userId]);
    echo json_encode(array('success' => true, 'amount' => $coupon['value']));
}

// VULN-33 [TIER-4]: ReDoS - User-controlled regex pattern
function validateWithPattern() {
    $pattern = $_POST['pattern'];
    $input = $_POST['input'];
    // User controls the regex — potential catastrophic backtracking
    if (preg_match('/' . $pattern . '/', $input)) { // nosec — ReDoS + regex injection
        echo json_encode(array('matched' => true));
    } else {
        echo json_encode(array('matched' => false));
    }
}

// VULN-34 [TIER-4]: CORS Misconfiguration - Reflected Origin
function setCorsHeaders() {
    $origin = isset($_SERVER['HTTP_ORIGIN']) ? $_SERVER['HTTP_ORIGIN'] : '*';
    header("Access-Control-Allow-Origin: " . $origin); // nosec — reflects any origin
    header("Access-Control-Allow-Credentials: true");
}

// VULN-35 [TIER-4]: Command Injection - Conditional branch, only one path vulnerable
function generateReport() {
    $format = $_POST['format'];
    $reportId = $_POST['id'];

    if ($format === 'pdf') {
        // Safe: uses escapeshellarg
        $safe = escapeshellarg($reportId);
        $output = array();
        $retval = 0;
        exec("wkhtmltopdf reports/{$safe}.html reports/{$safe}.pdf", $output, $retval);
    } else {
        // Vulnerable: no escaping — INTENTIONAL for scanner testing
        $output = array();
        $retval = 0;
        exec("cat reports/" . $reportId . "." . $format, $output, $retval); // nosec — intentional fixture
    }
}

// VULN-36 [TIER-4]: Encoded Injection - Base64 decoded then executed
function executeEncodedTask() {
    $encoded = $_POST['command'];
    $decoded = base64_decode($encoded); // nosec
    // Looks safe because it's "internal encoded format" — INTENTIONAL for scanner testing
    $output = array();
    $retval = 0;
    exec($decoded, $output, $retval); // nosec — intentional test fixture
    return $output;
}

// VULN-37 [TIER-4]: Mass Assignment - Unfiltered $_POST to database update
function updateProfile() {
    $userId = $_SESSION['user_id'];
    $updates = $_POST; // attacker includes: role=admin&is_verified=1
    $conn = new PDO("mysql:host=localhost;dbname=app", "root", "root");
    $setParts = array();
    $values = array();
    foreach ($updates as $key => $value) {
        $setParts[] = "{$key} = ?"; // nosec — column names from user input
        $values[] = $value;
    }
    $values[] = $userId;
    $setClause = implode(", ", $setParts);
    $conn->prepare("UPDATE users SET {$setClause} WHERE id = ?")->execute($values); // nosec
    echo "updated";
}

// VULN-38 [TIER-4]: Information Disclosure - phpinfo and error details
function debugEndpoint() {
    if (isset($_GET['debug'])) {
        phpinfo(); // nosec — exposes full server configuration
    }
}

function handleError($error) {
    global $config;
    echo json_encode(array(
        'error' => $error->getMessage(),
        'trace' => $error->getTraceAsString(), // nosec — full stack trace
        'file' => $error->getFile(),           // nosec — internal paths
        'line' => $error->getLine(),
        'server' => $_SERVER['SERVER_NAME'],   // nosec
        'db_host' => $config['database']['host'], // nosec
    ));
}

// VULN-39 [TIER-4]: Arbitrary File Write - Upload with unchecked filename
function uploadFile() {
    $file = $_FILES['file'];
    $filename = $file['name']; // attacker: "../../../var/www/shell.php"
    $dest = "/var/uploads/" . $filename; // nosec — no sanitization
    move_uploaded_file($file['tmp_name'], $dest); // nosec
    echo json_encode(array('saved' => $dest));
}

// VULN-40 [TIER-4]: Object Injection via JSON + dynamic method call
class UserAction {
    public $action;
    public $params;

    public function execute() {
        // Dynamic method invocation from deserialized data
        $method = $this->action;
        return $this->$method($this->params); // nosec — arbitrary method call
    }

    private function deleteAccount($params) {
        // destructive action
        $conn = new PDO("mysql:host=localhost;dbname=app", "root", "root");
        $conn->prepare("DELETE FROM users WHERE id = ?")->execute([$params['user_id']]);
    }

    private function escalatePrivileges($params) {
        $conn = new PDO("mysql:host=localhost;dbname=app", "root", "root");
        $conn->prepare("UPDATE users SET role = 'admin' WHERE id = ?")->execute([$params['user_id']]);
    }
}

function handleAction() {
    $data = json_decode(file_get_contents("php://input"), true);
    $action = new UserAction();
    $action->action = $data['action']; // attacker: "escalatePrivileges"
    $action->params = $data['params'];
    $action->execute(); // nosec — arbitrary method invocation
}

// Error reporting enabled — exposes errors to users
error_reporting(E_ALL); // nosec
ini_set('display_errors', '1'); // nosec

// Route handling
$route = $_SERVER['REQUEST_URI'];
$method = $_SERVER['REQUEST_METHOD'];

setCorsHeaders();

switch (true) {
    case $route === '/api/user': getUserById($_GET['id']); break;
    case $route === '/page': displayName(); break;
    case $route === '/ping': pingHost(); break;
    case $route === '/file': readFile(); break;
    case $route === '/calc': calculate(); break;
    case $route === '/redirect': doRedirect(); break;
    case $route === '/login': setSessionCookie(); break;
    case $route === '/api/search': searchProducts(); break;
    case $route === '/api/convert': convertImage(); break;
    case $route === '/profile': showProfile(); break;
    case $route === '/api/fetch': fetchUrl(); break;
    case $route === '/api/register': registerUser(); break;
    case $route === '/api/orders': getOrdersByStatus(); break;
    case $route === '/api/products': listProducts(); break;
    case $route === '/comment': showComment(); break;
    case $route === '/api/report': handleReportRequest(); break;
    case $route === '/api/health': checkServiceHealth(); break;
    case $route === '/api/xml': parseXml(); break;
    case $route === '/api/session': restoreSession(); break;
    case $route === '/auth/callback': authCallback(); break;
    case $route === '/api/coupon': redeemCoupon(); break;
    case $route === '/api/validate': validateWithPattern(); break;
    case $route === '/api/generate': generateReport(); break;
    case $route === '/api/profile': updateProfile(); break;
    case $route === '/api/upload': uploadFile(); break;
    case $route === '/api/action': handleAction(); break;
    case $route === '/debug': debugEndpoint(); break;
}
?>
